File: map.py
from pyray import *
import random

class Map:
    max_x = 100
    max_y = 10
    dirt = load_texture("asset/dirt.png")
    size = Vector2(16, 16)

    def __init__(self):
        self.map = [[0 for i in range(Map.max_x)] for j in range(Map.max_y)]
        self.center = Vector2(0, 0)
        for y in range(Map.max_y):
            for x in range(Map.max_x):
                self.map[y][x] = random.choice([0, 1])

    def collide(self, rect, direction, is_grounded):
        jump_stop = False
        for y in range(Map.max_y):
            for x in range(Map.max_x):
                if self.map[y][x] == 1:
                    rec = Rectangle(self.center.x + x * Map.size.x, self.center.y + y * Map.size.y, Map.size.x, Map.size.y)
                    if check_collision_recs(rec, rect):
                        if rect.x + rect.width > rec.x + rec.width and direction.x < 0:
                            rect.x = rec.x + rec.width
                        if rect.y + rect.height > rec.y + rec.height and direction.y < 0:
                            rect.y = rec.y + rec.height
                            jump_stop = True
                        if rect.x < rec.x and direction.x > 0:
                            rect.x = rec.x - rect.width
                        if rect.y < rec.y and direction.y > 0:
                            rect.y = rec.y - rect.height
                            is_grounded = True
        return (rect, is_grounded, jump_stop)

    def draw(self):
        for y in range(Map.max_y):
            for x in range(Map.max_x):
                if self.map[y][x] == 1:
                    rec = Rectangle(0, 0, Map.size.x, Map.size.y)
                    position = Vector2(self.center.x + x * Map.size.x, self.center.y + y * Map.size.y)
                    draw_texture_rec(Map.dirt, rec, position, WHITE)

File: map.go
package main

import (
	. "exp-raylib/shared"
	. "github.com/gen2brain/raylib-go/raylib"
)

var (
	MapLoaded   bool
	Map         [MaxY][MaxX]Block
	dirtTexture Texture2D
	texture_x   = i32(16)
	texture_y   = i32(16)
	offset_x    = -i32(MaxX) * texture_x / 2
	offset_y    = i32(0)
)

func MapInit() {
	dirtTexture = LoadTexture("asset/dirt.png")
}

func MapCollide(rec *Rectangle) bool {
	t := RectangleInt32{}
	t.Width = texture_x
	t.Height = texture_y
	center := RectCenter(*rec)
	cx, cy := MapIndex(center)
	n := 9
	for y := range MaxY {
		iy := y - n/2 + cy
		if !MapInsideY(iy) {
			continue
		}
		position_y := i32(iy)*texture_y + offset_y
		t.Y = position_y
		for x := range n {
			ix := x - n/2 + cx
			if !MapInsideX(ix) {
				continue
			}
			position_x := i32(ix)*texture_x + offset_x
			t.X = position_x
			if Map[iy][ix] == Dirt {
				tile := t.ToFloat32()
				if CheckCollisionRecs(tile, *rec) {
					return true
				}
			}
		}
	}
	return false
}

func RectCenter(r Rectangle) Vector2 {
	return NewVector2(r.X+r.Width*0.5, r.Y+r.Height*0.5)
}

func MapDraw() {
	cameraRect := CameraRect(0)
	rect := Rectangle{}
	rect.Width = f32(texture_x)
	rect.Height = f32(texture_y)
	center := RectCenter(cameraRect)
	cx, cy := MapIndex(center)
	n := 61
	for y := range n {
		iy := y - n/2 + cy
		if !MapInsideY(iy) {
			continue
		}
		position_y := i32(iy)*texture_y + offset_y
		rect.Y = f32(position_y)
		for x := range n {
			ix := x - n/2 + cx
			if !MapInsideX(ix) {
				continue
			}
			position_x := i32(ix)*texture_x + offset_x
			rect.X = f32(position_x)
			if Map[iy][ix] == Dirt {
				if CheckCollisionRecs(rect, cameraRect) {
					DrawTexture(dirtTexture, position_x, position_y, White)
				}
			}
		}
	}
}

func MapIndex(position Vector2) (int, int) {
	x := (i32(position.X) - offset_x) / texture_x
	y := (i32(position.Y) - offset_y) / texture_y
	return int(x), int(y)
}

func MapInsideX(x int) bool {
	return x >= 0 && x < MaxX
}

func MapInsideY(y int) bool {
	return y >= 0 && y < MaxY
}

func MapRect(x, y int) Rectangle {
	r := Rectangle{}
	r.X = f32(i32(x)*texture_x + offset_x)
	r.Y = f32(i32(y)*texture_y + offset_y)
	r.Width = f32(texture_x)
	r.Height = f32(texture_y)
	return r
}
